### PR TITLE
Introduced CcdCaseSubmitter and ExceptionRecordProvider

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
N/A

### Change description ###
Making CreateCaseCallbackService better maintainable - extracted ExceptionRecordProvider and CcdCaseSubmitter components. This is the second step of refactoring after using PaymentsProcessor. Potentially there will be one more step.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.ExceptionRecordFields.DISPLAY_WARNINGS;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.ExceptionRecordFields.OCR_DATA_VALIDATION_WARNINGS;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.YesNoFieldValues.NO;
+
+class ExceptionRecordProviderTest {
+
+    public static final long CASE = 100L;
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProviderTest.java]: is this an ID?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProvider.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.ExceptionRecordFields.CASE_REFERENCE;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.ExceptionRecordFields.DISPLAY_WARNINGS;
+import static uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd.definition.ExceptionRecordFields.OCR_DATA_VALIDATION_WARNINGS;
+
+@Service
+public class ExceptionRecordProvider {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProvider.java]: Sound very generic, while setting 'displayWarnings' to 'no' and clearing 'ocrDataValidationWarnings' is very specific and used in one particular scenario.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProvider.java]: same problem with `CcdCaseSubmitter`

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProviderTest.java (hunk 1 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProviderTest.java]: Renamed

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProvider.java (hunk 2 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordProvider.java]: Renamed

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java [hunk 3]
```diff
@@ -505,152 +455,8 @@ void should_report_errors_when_form_type_is_null() {
...
-
-        // then
-        assertThat(result.getErrors()).isEmpty();
-        assertThat(result.getWarnings()).isEmpty();
-
-        verify(paymentsProcessor).updatePayments(caseDetails, CASE_ID);
-    }
-
-    @Test
-    void should_return_error_if_publishing_payment_command_failed() throws Exception {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: looks like this test is gone

Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: Restored in CcdNewCaseCreatorTest

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreator.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+import uk.gov.hmcts.reform.ccd.client.model.Event;
+import uk.gov.hmcts.reform.ccd.client.model.StartEventResponse;
+
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
+
+@Service
+public class CcdNewCaseCreator {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreator.java]:
```suggestion
public class CcdCaseCreator {
```
Up to you

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreator.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+    @SuppressWarnings({"squid:S2139", "unchecked"}) // squid for exception handle + logging
+    public ProcessResult createNewCase(
+        ExceptionRecord exceptionRecord,
+        ServiceConfigItem configItem,
+        boolean ignoreWarnings,
+        String idamToken,
+        String userId,
+        CaseDetails exceptionRecordData
+    ) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreator.java]: This method is way too long. I realise it was like this in its original version and we should make improvements one step at a time, so this is fine for this PR. Perhaps something for the future, though.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java [hunk 6]
```diff
@@ -199,7 +173,7 @@ private ProcessResult tryCreateNewCase(
                 );
             } else if (ids.size() == 1) {
                 return new ProcessResult(
-                    prepareResultExceptionRecord(
+                    exceptionRecordFinalizer.finalizeExceptionRecord(
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Worth creating a local variable and set it with the result of `exceptionRecordFinalizer.finalizeExceptionRecord`. It wasn't clear to me what it returned. This is just a suggestion to consider.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java [hunk 7]
```diff
@@ -78,19 +60,13 @@
...
-    private CoreCaseDataApi coreCaseDataApi;
+    private CcdApi ccdApi;
 
     @Mock
-    private CcdApi ccdApi;
+    private CcdNewCaseCreator ccdNewCaseCreator;
 
     @Mock
-    private PaymentsProcessor paymentsProcessor;
+    private ExceptionRecordFinalizer exceptionRecordFinalizer;
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: Now that we have this ExceptionRecordFinalizer (which is a good idea), we should verify that it was called properly and also that we return its result.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        originalValues.put("field1", "value1");
+
+        // when
+        Map<String,Object> res = exceptionRecordFinalizer.finalizeExceptionRecord(originalValues, CASE_ID);
+
+        // then
+        assertThat(res.get("field1")).isEqualTo("value1");
+        assertThat(res.get(ExceptionRecordFields.CASE_REFERENCE)).isEqualTo(Long.toString(CASE_ID));
+        assertThat(res.get(DISPLAY_WARNINGS)).isEqualTo(NO);
+        assertThat(res.get(OCR_DATA_VALIDATION_WARNINGS)).isEqualTo(emptyList());
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java]: Worth adding an assertion that there are no more keys than those 4.

Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java]: Will do

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java (hunk 7 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: Will do

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java (hunk 6 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackService.java]: Will do

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreatorTest.java [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                    caseDetails
+                );
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getWarnings()).isEmpty();
+
+        verify(exceptionRecordFinalizer).finalizeExceptionRecord(anyMap(), anyLong());
+        verify(paymentsProcessor).updatePayments(caseDetails, CASE_ID);
+        verify(exceptionRecordFinalizer).finalizeExceptionRecord(anyMap(), anyLong());
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreatorTest.java]: This check is already present

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java [hunk 10]
```diff
@@ -314,54 +300,32 @@ void should_report_error_if_classification_is_supplementary_evidence() {
...
-            caseDetails(basicCaseData()),
+            caseDetails(caseData),
             true
         ), IDAM_TOKEN, USER_ID);
 
         // then
-        assertThat(result.getExceptionRecordData().get(CASE_REFERENCE)).isEqualTo("345");
-        assertThat(result.getWarnings().isEmpty()).isTrue();
-        assertThat(result.getErrors().isEmpty()).isTrue();
+        assertThat(result.getExceptionRecordData()).isEmpty();
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: Worth setting up the finaliser to return some simple non-empty map and this assertion would then verify it's the same map.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+        // when
+        Map<String,Object> res = exceptionRecordFinalizer.finalizeExceptionRecord(originalValues, CASE_ID);
+
+        // then
+        assertThat(res).containsOnlyKeys("field1", CASE_REFERENCE, DISPLAY_WARNINGS, OCR_DATA_VALIDATION_WARNINGS);
+        assertThat(res.get("field1")).isEqualTo("value1");
+        assertThat(res.get(CASE_REFERENCE)).isEqualTo(Long.toString(CASE_ID));
+        assertThat(res.get(DISPLAY_WARNINGS)).isEqualTo(NO);
+        assertThat(res.get(OCR_DATA_VALIDATION_WARNINGS)).isEqualTo(emptyList());
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java]:
```suggestion
        assertThat(res.get(OCR_DATA_VALIDATION_WARNINGS)).isEqualTo(emptyList());
        assertThat(res.size()).isEqualTo(4);
```

Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/ExceptionRecordFinalizerTest.java]: My bad, there's `containsOnlyKeys` already.

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java (hunk 10 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: Have done

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreatorTest.java (hunk 9 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdNewCaseCreatorTest.java]: Will fix

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java (hunk 10 above)
Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CreateCaseCallbackServiceTest.java]: where?

**Pull request merged**